BJS2-40812 Post Service: Comments feature

drewbal opened the pull request.

drewbal committed:
Feat: adds create comment dto and response dto

drewbal committed:
Feat: adds create comment endpoint

drewbal committed:
Feat: gets Post by id

PostService: getPostById method added

drewbal committed:
Feat: creates new comment

CommentService: create method added

drewbal committed:
Feat: validates post exists by id

PostValidator: validatePostExistsById method added

drewbal committed:
Feat: endpoint creates new comment

CommentController: create method added

drewbal committed:
Feat: minor fixes

Logs added, access modifiers fixed, syntax fixes

drewbal committed:
Other: adds MockMVC testing for create method in controller

YTXIRE commented on src/main/java/faang/school/postservice/controller/CommentController.java:
```diff
@@ -0,0 +1,30 @@
+    public ResponseEntity<CommentDto> create(
+            @PathVariable @Positive(message = "Post id should be a positive number") long postId,
+            @Valid @RequestBody CreateCommentDto dto) {
+        log.info("Request for new commit for post #{} from user #{}", postId, dto.getAuthorId());
```
Думаешь логирование в самом контроллере нужно? Не лучше его в вынести на уровень сервиса?

drewbal replied:
Я хочу сделать логирование на входе и выходе, т.е. отследить момент, что запрос пришел и метод начал работать, а также делаю логирование на выходе из сервиса, если операция завершается удачно.

Пока наверное это больше для себя, со временем разберусь с необходимым уровнем детализации логирования

YTXIRE commented on src/main/java/faang/school/postservice/dto/comment/CommentDto.java:
```diff
@@ -0,0 +1,19 @@
+
+import java.time.LocalDateTime;
+import java.util.List;
+
+@Builder
+@Data
+public class CommentDto {
+    private long id;
+    private String content;
+    private long authorId;
+    private List<Long> likeIds;
+    private Long postId;
+    private LocalDateTime createdAt;
+    private LocalDateTime updatedAt;
```
Не хвататет валидации полей

drewbal replied:
А она тут и не нужна :) Как я объяснял, это дто - формат исходящего ответа на запрос, когда мы предоставляем информацию пользователю. Учитывая, что это формируется напрямую из сущности, валидация не требуется.

YTXIRE replied:
Значит ее лучше поместить в папку responce, чтобы не было вопросов

drewbal replied:
Я добавлю в имя класса Response, чтобы не было путаницы

YTXIRE commented on src/main/java/faang/school/postservice/service/CommentService.java:
```diff
@@ -0,0 +1,36 @@
+    private final CommentMapper commentMapper;
+    private final PostValidator postValidator;
+    private final UserServiceClient userServiceClient;
+
```
Лишняя пустая строка

YTXIRE commented on src/main/java/faang/school/postservice/service/CommentService.java:
```diff
@@ -0,0 +1,36 @@
+
+    public CommentDto create(long postId, CreateCommentDto dto) {
+        postValidator.validatePostExistsById(postId);
+        userServiceClient.getUser(dto.getAuthorId());
```
А он работает?

drewbal replied:
Пока еще не тестил, а в юнит тестах мокаю его. Еще не разбирался как он именно работает, использую как уже готовый тул, предоставленный для задачи. 

Завершишь свою часть и можно будет играться, а то у меня все упирается в отсутствие постов, без них сильно не покомментишь :))

YTXIRE replied:
Там будет проблема с методами и тем, что не будет заводиться. Чуть позже закину отдельных пр.

YTXIRE commented on src/main/java/faang/school/postservice/controller/CommentController.java:
```diff
@@ -0,0 +1,30 @@
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequestMapping("/posts")
```
Опиши в описания мр, почему такая логика, чтобы вопросов не было.

YTXIRE commented on src/test/java/faang/school/postservice/validator/PostValidatorTest.java:
```diff
@@ -0,0 +1,45 @@
+        Exception ex = assertThrows(EntityNotFoundException.class, () -> postValidator.validatePostExistsById(postId));
+        assertEquals("Post with id #1 doesn't exist", ex.getMessage());
+    }
+
```
Лишняя пустая строка

drewbal committed:
Feat: adds Global Exception Handler

Other: final modifier for bean in CommentService added

drewbal committed:
Fix: Cross-Review fixes

Blank lines fixed

drewbal committed:
Feat: adds new UpdateCommentDto, new Exception

UpdateCommentDto added;
UnauthorizedAccessException added to validate comment's author;
Global Exception Handler updated

drewbal committed:
Feat: adds new endpoint updateComment

drewbal committed:
Feat: adds new validations for comments

ValidateAuthorExistsById added
Validate AuthorId added

drewbal committed:
Feat: adds update logic in service

CommentService: updateComment added

YTXIRE commented on src/main/java/faang/school/postservice/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,62 @@
+package faang.school.postservice.exception;
```
Я отдельным пр закинул exceptions, поэтому, как сольют подлей к себе

drewbal committed:
Feat: adds get all comments endpoint in controller

CommentController: getAllComments method added

drewbal committed:
Feat: adds get all comments method in service

CommentService: getAllComments method added
CommentMapper: toListDto method added

drewbal committed:
Feat: adds delete comment endpoint in controller

CommentController: deleteComment method added

drewbal committed:
Feat: adds delete comment method in service

CommentService: deleteComment method added

drewbal committed:
Merge branch 'Dragon-master-stream7' into BJS2-40812-PostService-Comments

# Conflicts:
#	src/main/java/faang/school/postservice/exception/GlobalExceptionHandler.java

drewbal committed:
Feat: adds delete comment method in service

CommentService: deleteComment method added

drewbal committed:
Fix: cross-review fixes

CommentDto changed to CommentResponseDto

denis-levkhov commented on src/main/java/faang/school/postservice/service/CommentService.java:
```diff
@@ -0,0 +1,82 @@
+        postValidator.validatePostExistsById(postId);
+        Post post = postService.getPostById(postId);
+        List<Comment> comments = post.getComments().stream()
+                .sorted(Comparator.comparing(Comment::getUpdatedAt, Comparator.nullsLast(Comparator.reverseOrder())))
```
Помещение null значений в конец списка - выглядит круто!

denis-levkhov commented on src/test/java/faang/school/postservice/controller/CommentControllerTest.java:
```diff
@@ -0,0 +1,198 @@
+
+        ResponseEntity<CommentResponseDto> result = commentController.create(postId, createDto);
+
+        mockMvc.perform(post("/posts/{postId}/comments", postId)
```
Новые фишки, работаем) Поспрашиваю как доберусь до тестов.

denis-levkhov reviewed (commented):
Шикарный код, очень круто

DenisIgnatenko commented on src/main/java/faang/school/postservice/controller/CommentController.java:
```diff
@@ -0,0 +1,58 @@
+    public ResponseEntity<CommentResponseDto> create(
+            @PathVariable @Positive(message = "Post id should be a positive number") long postId,
+            @Valid @RequestBody CreateCommentDto dto) {
+        log.info("Request for new commit for the post #{} from user #{}", postId, dto.getAuthorId());
```
Вот это супер, я в следующем сервисе тоже применю логирование. Должно быть удобно читать как обрабатываются запросы в консоли.

DenisIgnatenko commented on src/test/java/faang/school/postservice/controller/CommentControllerTest.java:
```diff
@@ -0,0 +1,198 @@
+    }
+
+    @Test
+    @DisplayName("Delete comment by id successfully")
```
ОООО, почему я сам не догадался таком именовании тестов? :)

RuslanZinovyev commented on src/main/java/faang/school/postservice/controller/CommentController.java:
```diff
@@ -0,0 +1,58 @@
+@RestController
+@RequestMapping("/posts")
+@RequiredArgsConstructor
+@Validated
```
Нам нужна эта аннотация здесь, ты используешь где то групповую валидацию? Я думаю ты можешь смело удалить эту аннотацию и валидация будет работатеь через Valid на увроне параметров метода. Если тебе эта часть не понятна прочитай внимательно статью что я поделился в knowledge-hub канале.

drewbal replied:
Она нам очень даже нужна. Дело в том, что @Valid работает только на объектах, передаваемых в метод, т.е. проверяет дто по аннотациям в ним. А вот для валидации параметров типа @PathVariable или @RequestParam она не подходит, требуется добавление @Validated на класс. Т.е. я не использую возможности групповой валидации, но должен валидировать переменные пути.

RuslanZinovyev replied:
Ты совершенно прав, я так зациклился на DTO что не заметил валидацию переменной в URL

RuslanZinovyev commented on src/main/java/faang/school/postservice/controller/CommentController.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    @PutMapping("/{postId}/comments")
+    public CommentResponseDto update(
```
Раз уж ты начал использовать ResponseEntity для одного ендпоинта имеет смысл использовать данный класс и для других, хорошая практика когда код однородный.

drewbal replied:
Да, наверное не лучшее решение. Решил сделать проще и не писать лишнего, раз по дефолту и так кидается нужный статус 200. Но согласен, что страдает единообразие. 

Пофиксил

RuslanZinovyev commented on src/main/java/faang/school/postservice/validator/PostValidator.java:
```diff
@@ -0,0 +1,18 @@
+
+    public void validatePostExistsById(Long postId) {
+        if (!postRepository.existsById(postId)) {
+            throw new EntityNotFoundException(String.format("Post with id #%s doesn't exist", postId));
```
Не совсем понимаю твою любовь добавлять # для числового идентификатора, немного усложняет чтение кода. Можешь не исправлять, просто это на самом деле не обязательное и даже спорное решение. Наличие id: 123 и так делает лог понятным.

drewbal replied:
Это скорее профдеформация, когда реквизиты документов (договоров, контрактов, законов и пр.) в тексте принято указывать через № или # :)) Поправлю и буду придерживаться предложенного тобой паттерна

RuslanZinovyev commented on src/main/java/faang/school/postservice/service/CommentService.java:
```diff
@@ -0,0 +1,82 @@
+
+    public CommentResponseDto createComment(long postId, CreateCommentDto dto) {
+        postValidator.validatePostExistsById(postId);
+        userServiceClient.getUser(dto.getAuthorId());
```
Что делает данный вызов?

drewbal replied:
Это метод межсервисного взаимодействия, должен возвращать юзера из UserService или кидать исключение, если юзер не найден. Я использовал этот метод, полагаясь на задание:

> В UserServiceClient:
> getUser(long id) — возвращает пользователя с данным id из user_service. Бросит исключение FeignException, > если такого пользователя нет в user_service.
> Эти методы можно использовать для валидации присутствия пользователя.

Не самое лучшее решение, потому что я сам еще не понимаю как работает FeignClient (записал изучить на досуге) и по сути доверился составителям задания. Тем временем Юра уже сделал эту часть в user_service

RuslanZinovyev replied:
FeignClient это REST client  - обычно используется чтобы отправить HTTP запроc.  Я просто не понимаю зачем ты делаешь этот запрос если ты не возвращаешь никакого юзера и ничего впоследствии с ним не делаешь, допустим исключения не было и запрос был сделан успешно, дальше что? Для чего это нужно? Просто для валидации что пользователь существует? Если так то наверное надо было отдельный метод валидации написать для этого. Пока оставим как есть.

RuslanZinovyev commented on src/main/java/faang/school/postservice/validator/CommentValidator.java:
```diff
@@ -0,0 +1,27 @@
+        }
+    }
+
+    public void validateCommentAuthorId(Comment comment, long AuthorId) {
```
переменные всегда начинаются с маленькой буквы, исправь на authorId

drewbal replied:
это гремлины сделали!

RuslanZinovyev commented on src/test/java/faang/school/postservice/controller/CommentControllerTest.java:
```diff
@@ -0,0 +1,198 @@
+                .andExpect(jsonPath("$[0].id").value(comments.get(0).getId()))
+                .andExpect(jsonPath("$[2].authorId").value(comments.get(2).getAuthorId()));
+
+        assertNotNull(commentService.getAllComments(postId));
```
```
List<CommentResponseDto> result = commentService.getAllComments(postId);
assertNotNull(result);
assertEquals(3, result.size());
```

drewbal replied:
А надо ли это, если используем MockMvc?

drewbal replied:
И будет ошибка, потому что метод вызывается дважды: 1) через .perform 2) через List result.

Мне кажется, что их вообще можно удалить

RuslanZinovyev commented on src/test/java/faang/school/postservice/service/CommentServiceTest.java:
```diff
@@ -0,0 +1,314 @@
+
+    @Test
+    @DisplayName("Get all comments success with one null date")
+    void testGetAllCommentsSuccessnullDate() {
```
`testGetAllCommentsSuccessWithNullDate`

RuslanZinovyev commented on src/test/java/faang/school/postservice/controller/CommentControllerTest.java:
```diff
@@ -0,0 +1,198 @@
+    }
+
+    @Test
+    @DisplayName("Create comment successfully")
```
success - придерживаемся одной стратегии для всех методов

drewbal replied:
Исправил все

RuslanZinovyev commented on src/test/java/faang/school/postservice/controller/CommentControllerTest.java:
```diff
@@ -0,0 +1,198 @@
+    void testCreateCommentSuccess() throws Exception {
+        when(commentService.createComment(postId, createDto)).thenReturn(responseDto);
+
+        ResponseEntity<CommentResponseDto> result = commentController.create(postId, createDto);
```
Тебе не нужен этот вызов здесь, запрос к АPI выполняет mockMvc тут
`mockMvc.perform(post("/posts/{postId}/comments", postId)`

Ты можешь переделать тест следующим образом:

```
@Test
@DisplayName("Create comment successfully")
void testCreateCommentSuccess() throws Exception {
    when(commentService.createComment(postId, createDto)).thenReturn(responseDto);

    mockMvc.perform(post("/posts/{postId}/comments", postId)
                    .contentType(MediaType.APPLICATION_JSON)
                    .content(objectMapper.writeValueAsString(createDto)))
            .andExpect(status().isCreated())
            .andExpect(jsonPath("$.id").value(responseDto.getId()))
            .andExpect(jsonPath("$.authorId").value(responseDto.getAuthorId()));

    verify(commentService, times(1)).createComment(postId, createDto);
}
```

drewbal replied:
Понял. Я читал, что MockMvc по сути подменяет собой вызов метода, как мы делали раньше, но получается дублировал функционал. Значит при использовании MockMvc проверяем только через него, ассерты не трогаем.

RuslanZinovyev replied:
Да mockMvc это фактически REST клиент, который шлет запросы на твой ендпоинт. 
andExpect методы делает асершн вызова

RuslanZinovyev commented on src/test/java/faang/school/postservice/controller/CommentControllerTest.java:
```diff
@@ -0,0 +1,198 @@
+        System.out.println(responseDto);
+        when(commentService.updateComment(postId, updateDto)).thenReturn(responseDto);
+
+        CommentResponseDto result = commentController.update(postId, updateDto);
```
этот вызов не нужен, ты дублируешь поведение mockMvc

RuslanZinovyev commented on src/test/java/faang/school/postservice/controller/CommentControllerTest.java:
```diff
@@ -0,0 +1,198 @@
+    @Test
+    @DisplayName("Update comment success")
+    void testUpdateCommentSuccess() throws Exception {
+        System.out.println(responseDto);
```
Ты забыл удалить эту строку

drewbal replied:
Упс, это была тестовая строка для поиска бага. Забыл удалить

RuslanZinovyev commented on src/test/java/faang/school/postservice/service/CommentServiceTest.java:
```diff
@@ -0,0 +1,314 @@
+    @Test
+    @DisplayName("Get all comments success")
+    void testGetAllCommentsSuccess() {
+        List<Comment> comments = List.of(
```
Лучше вынести эту логику в приватные вспомогательные методы и сделать тест более чистым.

drewbal committed:
Fix: review fixes

CommentController: ResponseEntity added where missed
Log messages fixed
Bad namings fixed
Tests fixed
